File: gke_customization/gke_catalog/report/slip_wise_diamond_detail_report/slip_wise_diamond_detail_report.py
# Copyright (c) 2025, Gurukrupa Export and contributors
# For license information, please see license.txt

import frappe
from frappe import _


def execute(filters=None):
    filters = filters or {}
    columns = get_columns()
    data = get_data(filters)
    return columns, data


def get_columns():
    return [
        {"fieldname": "action", "label": _("Action"), "fieldtype": "Data", "width": 120},
        {"fieldname": "main_slip", "label": _("Main Slip"), "fieldtype": "Link", "options": "Main Slip", "width": 220},
        {"fieldname": "creation_date", "label": _("Main Slip Date"), "fieldtype": "Date", "width": 110},
        {"fieldname": "metal_touch", "label": _("Metal Touch"), "fieldtype": "Data", "width": 100},
        {"fieldname": "company", "label": _("Company"), "fieldtype": "Link", "options": "Company", "width": 220, "hidden": 1},
        {"fieldname": "branch", "label": _("Branch"), "fieldtype": "Link", "options": "Branch", "width": 180, "hidden": 1},
        {"fieldname": "department", "label": _("Department"), "fieldtype": "Link", "options": "Department", "width": 180, "hidden": 1},
        {"fieldname": "manufacturer", "label": _("Manufacturer"), "fieldtype": "Link", "options": "Manufacturer", "width": 180, "hidden": 1},
        {"fieldname": "issue_wt", "label": _("Issue Wt"), "fieldtype": "Float", "precision": 3, "width": 110},
        {"fieldname": "receive_wt", "label": _("Receive Wt"), "fieldtype": "Float", "precision": 3, "width": 110},
        {"fieldname": "extra_issue_wt", "label": _("Extra Issue Wt"), "fieldtype": "Float", "precision": 3, "width": 130},
        {"fieldname": "extra_receive_wt", "label": _("Extra Receive Wt"), "fieldtype": "Float", "precision": 3, "width": 140},
        {"fieldname": "total_issue_wt", "label": _("Total Issue Wt"), "fieldtype": "Float", "precision": 3, "width": 130},
        {"fieldname": "total_receive_wt", "label": _("Total Receive Wt"), "fieldtype": "Float", "precision": 3, "width": 140},
    ]


def get_conditions(filters):
    conditions = []
    values = {}

    if filters.get("company"):
        conditions.append("ms.company = %(company)s")
        values["company"] = filters.get("company")

    if filters.get("branch"):
        conditions.append("mwo.branch = %(branch)s")
        values["branch"] = filters.get("branch")

    if filters.get("department"):
        conditions.append("ms.department = %(department)s")
        values["department"] = filters.get("department")

    if filters.get("manufacturer"):
        conditions.append("ms.manufacturer = %(manufacturer)s")
        values["manufacturer"] = filters.get("manufacturer")

    if filters.get("main_slip"):
        conditions.append("ms.name = %(main_slip)s")
        values["main_slip"] = filters.get("main_slip")

    if filters.get("from_date"):
        conditions.append("DATE(ms.creation) >= %(from_date)s")
        values["from_date"] = filters.get("from_date")

    if filters.get("to_date"):
        conditions.append("DATE(ms.creation) <= %(to_date)s")
        values["to_date"] = filters.get("to_date")

    return conditions, values


def get_data(filters):
    conditions, values = get_conditions(filters)
    where_clause = "WHERE " + " AND ".join(conditions) if conditions else ""

    query = f"""
        WITH eir_op_summary AS (
            SELECT
                eiro.manufacturing_work_order,
                SUM(CASE WHEN eir.type = 'Issue' THEN COALESCE(eiro.diamond_wt, 0) ELSE 0 END) AS issue_wt,
                SUM(CASE WHEN eir.type = 'Receive' THEN COALESCE(eiro.diamond_wt, 0) ELSE 0 END) AS receive_wt
            FROM `tabEmployee IR Operation` eiro
            INNER JOIN `tabEmployee IR` eir
                ON eir.name = eiro.parent
               AND eir.docstatus < 2
            WHERE COALESCE(eiro.diamond_wt, 0) > 0
            GROUP BY eiro.manufacturing_work_order
        ),
        slip_mwo AS (
            SELECT DISTINCT
                ms.name AS main_slip,
                DATE(ms.creation) AS creation_date,
                ms.metal_touch AS metal_touch,
                ms.company AS company,
                mwo.branch AS branch,
                ms.department AS department,
                ms.manufacturer AS manufacturer,
                mop.manufacturing_work_order
            FROM `tabMain Slip` ms
            LEFT JOIN `tabMain Slip Operation` mso
                ON mso.parent = ms.name
            LEFT JOIN `tabManufacturing Operation` mop
                ON mop.name = mso.manufacturing_operation
            LEFT JOIN `tabManufacturing Work Order` mwo
                ON mwo.name = mop.manufacturing_work_order
            {where_clause}
        ),
        ir_summary AS (
            SELECT
                sm.main_slip,
                sm.creation_date,
                sm.metal_touch,
                sm.company,
                MAX(sm.branch) AS branch,
                sm.department,
                sm.manufacturer,
                SUM(COALESCE(eos.issue_wt, 0)) AS issue_wt,
                SUM(COALESCE(eos.receive_wt, 0)) AS receive_wt
            FROM slip_mwo sm
            LEFT JOIN eir_op_summary eos
                ON eos.manufacturing_work_order = sm.manufacturing_work_order
            GROUP BY
                sm.main_slip,
                sm.creation_date,
                sm.metal_touch,
                sm.company,
                sm.department,
                sm.manufacturer
        ),
        extra_summary AS (
            SELECT
                ms.name AS main_slip,
                SUM(
                    CASE
                        WHEN COALESCE(msd.mop_qty, 0) > COALESCE(msd.mop_consume_qty, 0)
                        THEN COALESCE(msd.mop_qty, 0) - COALESCE(msd.mop_consume_qty, 0)
                        ELSE 0
                    END
                ) AS extra_issue_wt,
                SUM(
                    CASE
                        WHEN COALESCE(msd.mop_consume_qty, 0) > COALESCE(msd.mop_qty, 0)
                        THEN COALESCE(msd.mop_consume_qty, 0) - COALESCE(msd.mop_qty, 0)
                        ELSE 0
                    END
                ) AS extra_receive_wt
            FROM `tabMain Slip` ms
            LEFT JOIN `tabMain Slip SE Details` msd
                ON msd.parent = ms.name
            WHERE msd.variant_of = 'D'
            GROUP BY ms.name
        )
        SELECT
            'View Details' AS action,
            ir.main_slip,
            ir.creation_date,
            ir.metal_touch,
            ir.company,
            ir.branch,
            ir.department,
            ir.manufacturer,
            ROUND(COALESCE(ir.issue_wt, 0), 3) AS issue_wt,
            ROUND(COALESCE(ir.receive_wt, 0), 3) AS receive_wt,
            ROUND(COALESCE(ex.extra_issue_wt, 0), 3) AS extra_issue_wt,
            ROUND(COALESCE(ex.extra_receive_wt, 0), 3) AS extra_receive_wt,
            ROUND(COALESCE(ir.issue_wt, 0) + COALESCE(ex.extra_issue_wt, 0), 3) AS total_issue_wt,
            ROUND(COALESCE(ir.receive_wt, 0) + COALESCE(ex.extra_receive_wt, 0), 3) AS total_receive_wt
        FROM ir_summary ir
        LEFT JOIN extra_summary ex
            ON ex.main_slip = ir.main_slip
        WHERE
            COALESCE(ir.issue_wt, 0) > 0
            OR COALESCE(ir.receive_wt, 0) > 0
            OR COALESCE(ex.extra_issue_wt, 0) > 0
            OR COALESCE(ex.extra_receive_wt, 0) > 0
        ORDER BY
            ir.creation_date DESC,
            ir.main_slip
    """

    return frappe.db.sql(query, values, as_dict=True)


@frappe.whitelist()
def get_row_details(main_slip):
    details = frappe.db.sql("""
        WITH eir_detail AS (
            SELECT
                eiro.manufacturing_work_order,
                eir.employee,
                eir.type AS ir_type,
                ROUND(SUM(COALESCE(eiro.diamond_wt, 0)), 3) AS diamond_wt
            FROM `tabEmployee IR Operation` eiro
            INNER JOIN `tabEmployee IR` eir
                ON eir.name = eiro.parent
               AND eir.docstatus < 2
            WHERE COALESCE(eiro.diamond_wt, 0) > 0
            GROUP BY
                eiro.manufacturing_work_order,
                eir.employee,
                eir.type
        )
        SELECT
            ms.name AS main_slip,
            DATE(ms.creation) AS creation_date,
            ms.department AS department,
            ed.employee,
            mop.manufacturing_work_order AS batch_no,
            ed.ir_type,
            ed.diamond_wt
        FROM `tabMain Slip` ms
        INNER JOIN `tabMain Slip Operation` mso
            ON mso.parent = ms.name
        INNER JOIN `tabManufacturing Operation` mop
            ON mop.name = mso.manufacturing_operation
        INNER JOIN eir_detail ed
            ON ed.manufacturing_work_order = mop.manufacturing_work_order
        WHERE ms.name = %(main_slip)s
        ORDER BY
            ms.department,
            ed.employee,
            mop.manufacturing_work_order,
            ed.ir_type
    """, {
        "main_slip": main_slip
    }, as_dict=True)

    extra_details = frappe.db.sql("""
        SELECT
            msd.batch_no,
            ROUND(COALESCE(msd.mop_qty, 0), 3) AS mop_qty,
            ROUND(COALESCE(msd.mop_consume_qty, 0), 3) AS mop_consume_qty,
            ROUND(
                CASE
                    WHEN COALESCE(msd.mop_qty, 0) > COALESCE(msd.mop_consume_qty, 0)
                    THEN COALESCE(msd.mop_qty, 0) - COALESCE(msd.mop_consume_qty, 0)
                    ELSE 0
                END, 3
            ) AS extra_issue_wt,
            ROUND(
                CASE
                    WHEN COALESCE(msd.mop_consume_qty, 0) > COALESCE(msd.mop_qty, 0)
                    THEN COALESCE(msd.mop_consume_qty, 0) - COALESCE(msd.mop_qty, 0)
                    ELSE 0
                END, 3
            ) AS extra_receive_wt
        FROM `tabMain Slip SE Details` msd
        WHERE msd.parent = %(main_slip)s
          AND msd.variant_of = 'D'
          AND (
                COALESCE(msd.mop_qty, 0) > COALESCE(msd.mop_consume_qty, 0)
                OR COALESCE(msd.mop_consume_qty, 0) > COALESCE(msd.mop_qty, 0)
          )
        ORDER BY msd.batch_no
    """, {
        "main_slip": main_slip
    }, as_dict=True)

    return {
        "details": details,
        "extra_details": extra_details
    }

File: gke_customization/gke_catalog/report/slip_wise_diamond_detail_report/slip_wise_diamond_detail_report.js
frappe.query_reports["SLIP WISE DIAMOND DETAIL REPORT"] = {
    onload(report) {
        init_user_dept_permissions(report);

        $(document).off("click", ".view-details-btn");
        $(document).on("click", ".view-details-btn", function () {
            const main_slip = $(this).data("main-slip");
            const batch_no = $(this).data("batch-no");
            const employee = $(this).data("employee");

            frappe.call({
                method: "gke_customization.gke_catalog.report.slip_wise_diamond_detail_report.slip_wise_diamond_detail_report.get_row_details",
                args: {
                    main_slip: main_slip,
                    batch_no: batch_no,
                    employee: employee
                },
                callback: function (r) {
                    if (r.message) {
                        show_diamond_detail_popup(r.message, main_slip, batch_no, employee);
                    } else {
                        frappe.msgprint(__("No details found."));
                    }
                }
            });
        });
    },

    filters: [
        {
            fieldname: "company",
            label: __("Company"),
            fieldtype: "Link",
            options: "Company",
            reqd: 1,
            default: frappe.defaults.get_user_default("Company"),
            on_change: function () {
                frappe.query_report.set_filter_value("branch", "");
                frappe.query_report.set_filter_value("department", "");
            }
        },
        {
            fieldname: "branch",
            label: __("Branch"),
            fieldtype: "Link",
            options: "Branch",
            reqd: 0,
            get_query: function () {
                return {
                    filters: {
                        company: frappe.query_report.get_filter_value("company")
                    }
                };
            },
            on_change: function () {
                frappe.query_report.set_filter_value("department", "");
            }
        },
        {
            fieldname: "department",
            label: __("Department"),
            fieldtype: "Link",
            options: "Department",
            reqd: 1,
            get_query: function () {
                let company = frappe.query_report.get_filter_value("company");
                let branch = frappe.query_report.get_filter_value("branch");
                let filters = {};

                if (company) {
                    filters.company = company;
                }
                if (branch) {
                    filters.branch = branch;
                }

                return {
                    filters: filters
                };
            }
        },
        {
            fieldname: "manufacturer",
            label: __("Manufacturer"),
            fieldtype: "Link",
            options: "Manufacturer"
        },
        {
            fieldname: "main_slip",
            label: __("Main Slip"),
            fieldtype: "Link",
            options: "Main Slip"
        },
        {
            fieldname: "from_date",
            label: __("From Date"),
            fieldtype: "Date"
        },
        {
            fieldname: "to_date",
            label: __("To Date"),
            fieldtype: "Date"
        }
    ],

    formatter: function (value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);

        if (column.fieldname === "action" && data && data.main_slip) {
            return `<button class="btn btn-xs btn-primary view-details-btn"
                        data-main-slip="${frappe.utils.escape_html(data.main_slip || "")}"
                        data-batch-no="${frappe.utils.escape_html(data.batch_no || "")}"
                        data-employee="${frappe.utils.escape_html(data.employee || "")}">
                        <i class="fa fa-eye"></i> View Details
                    </button>`;
        }

        if (column.fieldtype === "Float" && value !== null && value !== undefined && value !== "") {
            let num = parseFloat(value);
            if (!isNaN(num)) {
                return num.toFixed(3);
            }
        }

        return value;
    }
};


function init_user_dept_permissions(report) {
    frappe.call({
        method: "frappe.client.get",
        args: {
            doctype: "User",
            name: frappe.session.user
        },
        callback: function (user_res) {
            const roles = ((user_res.message && user_res.message.roles) || []).map(r => r.role);
            const management_roles = [
                "Director",
                "CEO",
                "System Manager",
                "Branch Manager",
                "Department Manager"
            ];
            const is_management = roles.some(role => management_roles.includes(role));

            frappe.call({
                method: "frappe.client.get_value",
                args: {
                    doctype: "Employee",
                    filters: { user_id: frappe.session.user },
                    fieldname: ["company", "branch", "department"]
                },
                callback: function (r) {
                    if (!r.message) return;

                    if (r.message.company && report.get_filter("company")) {
                        report.get_filter("company").set_value(r.message.company);
                        if (!is_management) {
                            report.get_filter("company").df.read_only = 1;
                            report.get_filter("company").refresh();
                        }
                    }

                    if (r.message.branch && report.get_filter("branch")) {
                        report.get_filter("branch").set_value(r.message.branch);
                        if (!is_management) {
                            report.get_filter("branch").df.read_only = 1;
                            report.get_filter("branch").refresh();
                        }
                    }

                    if (r.message.department && report.get_filter("department")) {
                        report.get_filter("department").set_value(r.message.department);
                        if (!is_management) {
                            report.get_filter("department").df.read_only = 1;
                            report.get_filter("department").refresh();
                        }
                    }

                    setTimeout(() => report.refresh(), 500);
                }
            });
        }
    });
}


function show_diamond_detail_popup(data, main_slip, batch_no, employee) {
    let detail_rows = "";
    let extra_rows = "";

    (data.details || []).forEach(function (row) {
        detail_rows += `
            <tr>
                <td>${row.creation_date || ""}</td>
                <td>${row.branch || ""}</td>
                <td>${row.department || ""}</td>
                <td>${row.employee || ""}</td>
                <td>${row.batch_no || ""}</td>
                <td>${row.ir_type || ""}</td>
                <td style="text-align:right;">${format_float(row.diamond_wt)}</td>
            </tr>
        `;
    });

    if (!detail_rows) {
        detail_rows = `
            <tr>
                <td colspan="7" style="text-align:center; color:#777;">No issue/receive details found</td>
            </tr>
        `;
    }

    (data.extra_details || []).forEach(function (row) {
        extra_rows += `
            <tr>
                <td>${row.batch_no || ""}</td>
                <td style="text-align:right;">${format_float(row.mop_qty)}</td>
                <td style="text-align:right;">${format_float(row.mop_consume_qty)}</td>
                <td style="text-align:right;">${format_float(row.extra_issue_wt)}</td>
                <td style="text-align:right;">${format_float(row.extra_receive_wt)}</td>
            </tr>
        `;
    });

    if (!extra_rows) {
        extra_rows = `
            <tr>
                <td colspan="5" style="text-align:center; color:#777;">No extra issue/receive details found</td>
            </tr>
        `;
    }

    const html = `
        <div>
            <div style="margin-bottom: 15px;">
                <p><strong>Main Slip:</strong> ${main_slip || ""}</p>
                <p><strong>Batch No:</strong> ${batch_no || ""}</p>
                <p><strong>Employee:</strong> ${employee || ""}</p>
            </div>

            <h5 style="margin-top: 10px;">Employee IR Details</h5>
            <div style="overflow-x:auto; margin-bottom:20px;">
                <table class="table table-bordered">
                    <thead>
                        <tr>
                            <th>Main Slip Date</th>
                            <th>Branch</th>
                            <th>Department</th>
                            <th>Employee</th>
                            <th>Batch No / MWO</th>
                            <th>Type</th>
                            <th style="text-align:right;">Diamond Wt</th>
                        </tr>
                    </thead>
                    <tbody>
                        ${detail_rows}
                    </tbody>
                </table>
            </div>

            <h5 style="margin-top: 10px;">Extra Issue / Receive Details</h5>
            <div style="overflow-x:auto;">
                <table class="table table-bordered">
                    <thead>
                        <tr>
                            <th>Batch No</th>
                            <th style="text-align:right;">MOP Qty</th>
                            <th style="text-align:right;">MOP Consume Qty</th>
                            <th style="text-align:right;">Extra Issue Wt</th>
                            <th style="text-align:right;">Extra Receive Wt</th>
                        </tr>
                    </thead>
                    <tbody>
                        ${extra_rows}
                    </tbody>
                </table>
            </div>
        </div>
    `;

    let d = new frappe.ui.Dialog({
        title: __("Diamond Detail View"),
        size: "extra-large",
        fields: [
            {
                fieldtype: "HTML",
                fieldname: "details_html"
            }
        ]
    });

    d.fields_dict.details_html.$wrapper.html(html);
    d.show();
}


function format_float(val) {
    let num = parseFloat(val || 0);
    return isNaN(num) ? "0.000" : num.toFixed(3);
}